File: robot.c
#include <stdio.h>

#define MAXN 1000

int a[MAXN][MAXN];
int dp[MAXN][MAXN];
int parent_i[MAXN][MAXN];
int parent_j[MAXN][MAXN];

int main() {
    int n;
    scanf("%d", &n);

    // Read matrix
    for(int i = 0; i < n; i++) {
        for(int j = 0; j < n; j++) {
            scanf("%d", &a[i][j]);
        }
    }

    // DP start
    dp[0][0] = a[0][0];
    parent_i[0][0] = -1;
    parent_j[0][0] = -1;

    // First row
    for(int j = 1; j < n; j++) {
        dp[0][j] = dp[0][j-1] + a[0][j];
        parent_i[0][j] = 0;
        parent_j[0][j] = j-1;
    }

    // First column
    for(int i = 1; i < n; i++) {
        dp[i][0] = dp[i-1][0] + a[i][0];
        parent_i[i][0] = i-1;
        parent_j[i][0] = 0;
    }

    // Fill rest
    for(int i = 1; i < n; i++) {
        for(int j = 1; j < n; j++) {
            if(dp[i-1][j] > dp[i][j-1]) {
                dp[i][j] = dp[i-1][j] + a[i][j];
                parent_i[i][j] = i-1;
                parent_j[i][j] = j;
            } else {
                dp[i][j] = dp[i][j-1] + a[i][j];
                parent_i[i][j] = i;
                parent_j[i][j] = j-1;
            }
        }
    }

    printf("Max value: %d\n", dp[n-1][n-1]);

    // Reconstruct path
    int mark[MAXN][MAXN] = {0};
    int i = n-1, j = n-1;

    while(i != -1 && j != -1) {
        mark[i][j] = 1;
        int pi = parent_i[i][j];
        int pj = parent_j[i][j];
        i = pi;
        j = pj;
    }

    // Print matrix with path
    printf("Path:\n");
    for(int i = 0; i < n; i++) {
        for(int j = 0; j < n; j++) {
            if(mark[i][j])
                printf("*%2d ", a[i][j]);
            else
                printf(" %2d ", a[i][j]);
        }
        printf("\n");
    }

    return 0;
}

File: Robot/robot.py
def dp_solution_with_path(a, n):
    dp = [[0]*n for _ in range(n)]
    parent = [[None]*n for _ in range(n)]

    dp[0][0] = a[0][0]

    # First row
    for j in range(1, n):
        dp[0][j] = dp[0][j-1] + a[0][j]
        parent[0][j] = (0, j-1)

    # First column
    for i in range(1, n):
        dp[i][0] = dp[i-1][0] + a[i][0]
        parent[i][0] = (i-1, 0)

    # Fill DP
    for i in range(1, n):
        for j in range(1, n):
            if dp[i-1][j] > dp[i][j-1]:
                dp[i][j] = dp[i-1][j] + a[i][j]
                parent[i][j] = (i-1, j)
            else:
                dp[i][j] = dp[i][j-1] + a[i][j]
                parent[i][j] = (i, j-1)

    # Reconstruct path
    path = set()
    i, j = n-1, n-1
    while True:
        path.add((i, j))
        if parent[i][j] is None:
            break
        i, j = parent[i][j]

    return dp[n-1][n-1], path


# INPUT (only once!)
n = int(input())
a = [list(map(int, input().split())) for _ in range(n)]

max_value, path = dp_solution_with_path(a, n)

print("Max value:", max_value)
print("Path:")

for i in range(n):
    for j in range(n):
        if (i, j) in path:
            print(f"*{a[i][j]:2}", end=" ")
        else:
            print(f" {a[i][j]:2}", end=" ")
    print()